compiler/rustc_codegen_llvm/src/type_of.rs

```diff
@@ -215,20 +215,16 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyAndLayout<'tcx> {
     /// of that field's type - this is useful for taking the address of
     /// that field and ensuring the struct has the right alignment.
     fn llvm_type<'a>(&self, cx: &CodegenCx<'a, 'tcx>) -> &'a Type {
+        // This must produce the same result for `repr(transparent)` wrappers as for the inner type!
+        // In other words, this should generally not look at the type at all, but only at the
+        // layout.
```

Comment:
Sorry for seeing this so late, but, if not made clear so far by anyone else:
the special cases you removed are *simultaneously* sound *and* obsolete.

They're sound because they only affect pointer types' pointee types (which doesn't impact ABI on it own), and they're obsolete thanks to LLVM opaque pointers (making the removal welcome!).

You can kinda tell they were working fine because e.g. `Option<&String>` would look like `i64*`, whereas the inner `&String` would actually be `{i8*, i64, i64}*` - worst case, it would pessimize some badly written LLVM optimizations, but all of those concerns should be gone now.

In fact, some of this code can very likely be simplified!

`scalar_pair_element_llvm_type` only exists in the form it does because it needed to special-case wide pointers, but now each component of the scalar pair can be independently converted to LLVM.

It might even make sense to remove caches like `scalar_lltypes` and instead have something much simpler keyed by `Primitive` - the existing `scalar_llvm_type_at` might be good enough, even, and I would argue the cache *only* existed because of the pointer cases, and can now be fully removed.

Reply:
> the special cases you removed are simultaneously sound and obsolete.

Yes that was my conclusion, too. That still makes this PR a good cleanup, no? :)

But looking at these special cases lead me to ones that aren't sound (see the recent [ABI issues](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AA-abi) I opened), so the cleanup of obsolete-but-sound type matching was still worth it.

I'm sure there's further simplification possible.